File: orchestrator/context.py
"""Context helpers for resolving paths and instance metadata."""

from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from typing import Any

from .inventory import resolve_relative


@dataclass(frozen=True)
class InstanceContext:
    id: str
    inventory_path: Path
    inventory_dir: Path
    generated_dir: Path
    config_dir: Path
    state_dir: Path
    workspace_root: Path
    gateway_port: int
    gateway_bind: str


CONTAINER_CONFIG_DIR = Path("/var/lib/openclaw/config")
CONTAINER_STATE_DIR = Path("/var/lib/openclaw/state")
CONTAINER_WORKSPACE_DIR = Path("/var/lib/openclaw/workspaces")


def build_instance_context(instance: dict[str, Any], inv_path: Path) -> InstanceContext:
    inv_dir = inv_path.parent
    paths = instance["paths"]
    host = instance["host"]

    return InstanceContext(
        id=str(instance["id"]),
        inventory_path=inv_path,
        inventory_dir=inv_dir,
        generated_dir=resolve_relative(inv_dir, paths["generated_dir"]),
        config_dir=resolve_relative(inv_dir, paths["config_dir"]),
        state_dir=resolve_relative(inv_dir, paths["state_dir"]),
        workspace_root=resolve_relative(inv_dir, paths["workspace_root"]),
        gateway_port=int(host["gateway_port"]),
        gateway_bind=str(host.get("bind", "127.0.0.1")),
    )

File: src/context.ts
import { resolve } from 'node:path';
import { InstanceContext } from './types';

export const CONTAINER_CONFIG_DIR = '/var/lib/openclaw/config';
export const CONTAINER_STATE_DIR = '/var/lib/openclaw/state';
export const CONTAINER_WORKSPACE_DIR = '/var/lib/openclaw/workspaces';

export function buildInstanceContext(
  instance: Record<string, unknown>,
  invPath: string,
): InstanceContext {
  const invDir = resolve(invPath, '..');
  const paths = record(instance.paths);
  const host = record(instance.host);

  const id = String(instance.id);

  return {
    id,
    inventoryPath: invPath,
    inventoryDir: invDir,
    generatedDir: resolve(invDir, String(paths.generated_dir ?? '.generated')),
    configDir: resolve(invDir, String(paths.config_dir ?? `instances/${id}/config`)),
    stateDir: resolve(invDir, String(paths.state_dir ?? `instances/${id}/state`)),
    workspaceRoot: resolve(invDir, String(paths.workspace_root ?? `instances/${id}/workspaces`)),
    gatewayPort: Number(host.gateway_port ?? 19789),
    gatewayBind: normalizeDockerBind(String(host.bind ?? '127.0.0.1')),
  };
}

function normalizeDockerBind(value: string): string {
  const normalized = value.trim().toLowerCase();

  if (!normalized || normalized === 'loopback' || normalized === 'localhost') {
    return '127.0.0.1';
  }

  if (normalized === 'all') {
    return '0.0.0.0';
  }

  return value;
}

function record(value: unknown): Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value)
    ? (value as Record<string, unknown>)
    : {};
}
